fetcher/fetcher.go

```diff
@@ -0,0 +1,258 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fetcher
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+
+	"github.com/ava-labs/hypersdk/keys"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+// Fetcher retrieves values on-the-fly and ensures that
+// a value is only fetched from disk once. Subsequent
+// requests can be retrieved from cache.
+type Fetcher struct {
+	im        state.Immutable
+	cacheLock sync.RWMutex
+	cache     map[string]*fetchData
+
+	keysToFetch map[string]*keyData        // Number of txns waiting for a key
+	txnsToFetch map[ids.ID]*sync.WaitGroup // Number of keys a txn is waiting on
+	keyLock     sync.RWMutex
+	txnLock     sync.Mutex
+
+	stopOnce  sync.Once
+	stop      chan struct{}
+	err       error
+	fetchable chan *task
+	l         sync.Mutex
+	wg        sync.WaitGroup
+
+	completed int
+	numTxs    int
+}
+
+// Data to insert into the cache
+type fetchData struct {
+	v      []byte
+	exists bool
+	chunks uint16
+}
+
+// Data pertaining to whether a key was fetched or which txns are waiting
+type keyData struct {
+	fetched bool
+	queue   []ids.ID
+}
+
+// Holds the information that a worker needs to fetch values
+type task struct {
+	ctx context.Context
+	id  ids.ID
+	key string
+}
+
+// New creates a new [Fetcher]
+func New(numTxs int, concurrency int, im state.Immutable) *Fetcher {
+	f := &Fetcher{
+		im:          im,
+		cache:       make(map[string]*fetchData, numTxs),
+		keysToFetch: make(map[string]*keyData),
+		txnsToFetch: make(map[ids.ID]*sync.WaitGroup, numTxs),
+		fetchable:   make(chan *task),
```

Comment:
Interesting choice to make this a non-buffered channel. I don't think we should go this route (as it could really impact the concurrency of execution if `Lookup` is blocking).